### ISSUE TITLE
Login on Firefox Android hangs forever on "Hashing your password, ..."

### ISSUE THREAD

**Issue by mkalinski** [xwiki-labs/cryptpad#602]
**Describe the bug**
After inputting login and password on https://cryptpad.fr/login/ and clicking "Log in", the splash page with spinner and text "Hashing your password, ..." is displayed forever.

**Where did it happen?**
CryptPad.fr

**To Reproduce**
1. Go to https://cryptpad.fr/
2. Click on "Log in"
3. Input username and password
4. Click on "Log in" button below the password
5. Splash page is displayed forever

**Expected behavior**
Splash page should go away and display my crypt drive

**Browser (please complete the following information):**
 - Device: Fairphone 2
 - OS: Android 7.1.2
 - Browser: Firefox Nightly
 - Version: 79.0.1
 - Extensions installed: UBlock Origin (cryptpad.fr whitelisted), Privacy Possum, Privacy Badger, HTTPS Everywhere, Dark Reader (cryptpad.fr disabled)
 - Browser tweaks: Firefox "Enhanced Tracking Protection" custom:
   - cookies: cross-site and social media trackers
   - tracking content: in all tabs
   - cryptominers: on
   - fingerprinters: on

**Additional context**
Anonymous crypt pads seem to be working.

I'm able to log in normally to my drive from desktop browser.

This bug also happened on previous versions of Firefox for Android (ex. ESR 68). I missed the exact point in time when this bug started occurring, but I was able to log in to crypt drive on this device with Firefox Android in the past.

Web Console output is as follows. The first CSP message comes from Privacy Badger failing to inject script.

```
13:14:17.284 Content Security Policy: The page’s settings blocked the loading of a resource at inline (“script-src”). 2 utils.js:35:9
13:14:19.696 Compiling [/customize/src/less2/pages/page-login.less] took 31ms LessLoader.js:161:25
13:14:20.421 asm.js type error: Disabled by debugger chainpad.dist.js
13:15:29.790 Script terminated by timeout at:
j@https://cryptpad.fr/bower_components/scrypt-async/scrypt-async.min.js?ver=3.21.0-0:6:2501
scrypt@https://cryptpad.fr/bower_components/scrypt-async/scrypt-async.min.js?ver=3.21.0-0:6:5924
factory/Cred.deriveFromPassphrase@https://cryptpad.fr/common/common-credential.js?ver=3.21.0-0:40:15
Exports.loginOrRegister/<@https://cryptpad.fr/customize/login.js?ver=3.21.0-0:151:18
nThen@https://cryptpad.fr/bower_components/nthen/index.js?ver=3.21.0-0:37:25
nThen@https://cryptpad.fr/bower_components/nthen/index.js?ver=3.21.0-0:67:16
Exports.loginOrRegister@https://cryptpad.fr/customize/login.js?ver=3.21.0-0:148:14
Exports.loginOrRegisterUI/</<@https://cryptpad.fr/customize/login.js?ver=3.21.0-0:456:25
setTimeout handler*Exports.loginOrRegisterUI/<@https://cryptpad.fr/customize/login.js?ver=3.21.0-0:455:20
setTimeout handler*Exports.loginOrRegisterUI@https://cryptpad.fr/customize/login.js?ver=3.21.0-0:447:16
@https://cryptpad.fr/login/main.js?ver=3.21.0-0:61:19
dispatch@https://cryptpad.fr/bower_components/jquery/dist/jquery.min.js?ver=3.21.0-0:3:7537
add/r.handle@https://cryptpad.fr/bower_components/jquery/dist/jquery.min.js?ver=3.21.0-0:3:5620
EventListener.handleEvent*add@https://cryptpad.fr/bower_components/jquery/dist/jquery.min.js?ver=3.21.0-0:3:6087
ja/<@https://cryptpad.fr/bower_components/jquery/dist/jquery.min.js?ver=3.21.0-0:3:5305
each@https://cryptpad.fr/bower_components/jquery/dist/jquery.min.js?ver=3.21.0-0:2:2861
each@https://cryptpad.fr/bower_components/jquery/dist/jquery.min.js?ver=3.21.0-0:2:845
ja@https://cryptpad.fr/bower_components/jquery/dist/jquery.min.js?ver=3.21.0-0:3:5281
on@https://cryptpad.fr/bower_components/jquery/dist/jquery.min.js?ver=3.21.0-0:3:11200
n.fn[b]@https://cryptpad.fr/bower_components/jquery/dist/jquery.min.js?ver=3.21.0-0:4:5724
@https://cryptpad.fr/login/main.js?ver=3.21.0-0:57:27
i@https://cryptpad.fr/bower_components/jquery/dist/jquery.min.js?ver=3.21.0-0:2:27151
add@https://cryptpad.fr/bower_components/jquery/dist/jquery.min.js?ver=3.21.0-0:2:27450
n.fn.ready@https://cryptpad.fr/bower_components/jquery/dist/jquery.min.js?ver=3.21.0-0:2:29515
n.fn.init@https://cryptpad.fr/bower_components/jquery/dist/jquery.min.js?ver=3.21.0-0:2:24945
n@https://cryptpad.fr/bower_components/jquery/dist/jquery.min.js?ver=3.21.0-0:2:405
@https://cryptpad.fr/login/main.js?ver=3.21.0-0:13:6
execCb@https://cryptpad.fr/bower_components/requirejs/require.js?ver=2.3.5:1696:33
check@https://cryptpad.fr/bower_components/requirejs/require.js?ver=2.3.5:883:51
enable/</<@https://cryptpad.fr/bower_components/requirejs/require.js?ver=2.3.5:1139:34
bind/<@https://cryptpad.fr/bower_components/requirejs/require.js?ver=2.3.5:134:23
emit/<@https://cryptpad.fr/bower_components/requirejs/require.js?ver=2.3.5:1189:23
each@https://cryptpad.fr/bower_components/requirejs/require.js?ver=2.3.5:59:31
emit@https://cryptpad.fr/bower_components/requirejs/require.js?ver=2.3.5:1188:21
check@https://cryptpad.fr/bower_components/requirejs/require.js?ver=2.3.5:938:30
enable/</<@https://cryptpad.fr/bower_components/requirejs/require.js?ver=2.3.5:1139:34
bind/<@https://cryptpad.fr/bower_components/requirejs/require.js?ver=2.3.5:134:23
emit/<@https://cryptpad.fr/bower_components/requirejs/require.js?ver=2.3.5:1189:23
each@https://cryptpad.fr/bower_components/requirejs/require.js?ver=2.3.5:59:31
emit@https://cryptpad.fr/bower_components/requirejs/require.js?ver=2.3.5:1188:21
check@https://cryptpad.fr/bower_components/requirejs/require.js?ver=2.3.5:938:30
enable/</<@https://cryptpad.fr/bower_components/requirejs/require.js?ver=2.3.5:1139:34
bind/<@https://cryptpad.fr/bower_components/requirejs/require.js?ver=2.3.5:134:23
scrypt-async.min.js:6:2501

```

**Comment by ansuz** [comment]
I think the problem stems from your custom settings for tracking protection. The system we use to establish a persistent session when you log in is indistinguishable from a tracking token since your browser needs to remember its own encryption keys to be able to access your content. These settings interfere with storage and can cause other random parts of CryptPad to fail.

Similarly, a lot of fingerprinting techniques rely on drawing invisible content using the HTML5 canvas element to pinpoint rendering quirks in different hardware. We use invisible canvases for legitimate purposes like creating thumbnails of higher resolution (encrypted) images on the client where a server would normally perform that task.

We can try to come up with workarounds to provide the same functionality without using these tools, but it's something of an arms race because the entities writing advanced trackers are also trying to work around the same limitations with malicious intent. Honest players are almost always at a disadvantage in these situations :frowning_face:

Anyway, you can try logging in after disabling those custom protections so we at least know for sure where the problem is coming from. The best case scenario might be that we detect when the failure occurs for this reason and inform the user what they can do to address the cause in their browser settings.

**Comment by mkalinski** [comment]
I disabled tracking protection and the Privacy Possum addon, and was able to log in successfully.

However, I then cleared cookies, enabled tracking protection and the addon, and then I was also able to log in.

There have been a couple of updates to firefox since I opened this issue. Either they resolved some underlying problem, or logging in once with tracking protection disabled changed something (I'm not sure if it's possible).

Either way, I can no longer reproduce this issue on my phone. It's unfortunate I wasn't able to confirm what the cause was.
